crates/symbolicator/src/endpoints/symbolicate_v2.rs

```diff
@@ -0,0 +1,194 @@
+use std::fs::File;
+use std::pin::pin;
+use std::sync::Arc;
+
+use axum::extract;
+use axum::response::Json;
+use futures::TryStreamExt;
+use serde::Deserialize;
+
+use symbolicator_native::interface::{ProcessMinidump, RewriteRules};
+use symbolicator_service::config::Config;
+use symbolicator_service::download::DownloadService;
+use symbolicator_service::types::Platform;
+use symbolicator_sources::SourceConfig;
+use tempfile::NamedTempFile;
+use tokio::io::{AsyncWriteExt, BufWriter};
+use tokio_util::io::StreamReader;
+
+use crate::service::{
+    RequestOptions, RequestQueryParams, RequestService, ScrapingConfig, SymbolicationResponse,
+};
+use crate::utils::sentry::ConfigureScope;
+
+use super::ResponseError;
+
+/// JSON body of the symbolication request.
+///
+/// This includes a bunch of common request options across the various symbolication requests.
+#[derive(Deserialize)]
+pub struct SymbolicationV2RequestBody {
+    pub platform: Option<Platform>,
+    #[serde(default)]
+    pub options: RequestOptions,
+    #[serde(default)]
+    pub scraping: ScrapingConfig,
+    #[serde(default)]
+    pub sources: Arc<[SourceConfig]>,
+
+    pub symbolicate: SymbolicationRequest,
+}
+
+/// The actual symbolication request.
+///
+/// This currently includes minidump and apple-crashreport based on stored attachments.
+/// In the future, we could move all of native, js and jvm symbolication to this enum.
+#[derive(Deserialize)]
+#[serde(tag = "type", rename_all = "lowercase")]
+pub enum SymbolicationRequest {
+    Minidump {
+        storage_url: String,
+        rewrite_first_module: RewriteRules,
+    },
+    AppleCrashreport {
+        storage_url: String,
+    },
+    // TODO: it should be possible to also support native, js and jvm requests here as well
+}
+
+pub async fn symbolicate_v2(
+    extract::State(service): extract::State<RequestService>,
+    extract::Query(params): extract::Query<RequestQueryParams>,
+    extract::Json(body): extract::Json<SymbolicationV2RequestBody>,
+) -> Result<Json<SymbolicationResponse>, ResponseError> {
+    sentry::start_session();
+
+    params.configure_scope();
+
+    let request_id = match body.symbolicate {
+        SymbolicationRequest::Minidump {
+            storage_url,
+            rewrite_first_module,
+        } => {
+            let minidump_file = make_minidump_file(service.config())?;
+            let (file, minidump_file) = minidump_file.into_parts();
+            let _ = download_attachment(service.download_svc(), &storage_url, file).await?;
```

Comment:
So this actually downloads the file as part of the request-response cycle, instead of async as part of the symbolication request?

Reply:
Indeed it does. Primarily so I don’t have to change the downstream `process_minidump` etc functions.
Though its also reasonable to move the downloading into the symbolication function itself. Maybe combined with a refactoring to remove the `TempPath` vs `File` distinction.